engine/engine.go

```diff
@@ -98,9 +114,170 @@ func ContextWithEngine(ctx context.Context) context.Context {
 	if !IsEngineEnabled() {
 		return ctx
 	}
-	return context.WithValue(ctx, TerraformCommandContextKey, &sync.Map{})
+	ctx = context.WithValue(ctx, TerraformCommandContextKey, &sync.Map{})
+	ctx = context.WithValue(ctx, LocksContextKey, util.NewKeyLocks())
+	return ctx
+}
+
+func DownloadEngine(ctx context.Context, opts *options.TerragruntOptions) error {
+	if !IsEngineEnabled() {
+		return nil
+	}
+	e := opts.Engine
+
+	if util.FileExists(e.Source) {
+		return nil
+	}
+	path, err := enginePath(e)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	if err := util.EnsureDirectory(path); err != nil {
+		return errors.WithStackTrace(err)
+	}
+	localEngineFile := filepath.Join(path, engineFileName(e))
+	if util.FileExists(localEngineFile) {
+		return nil
+	}
+	downloadFile := filepath.Join(path, enginePackageName(e))
+	var downloadURL string
+	if strings.HasPrefix(e.Source, "http") {
+		// if source starts with absolute path, download as is
+		downloadURL = e.Source
+	} else {
+		// Archive support documented in https://github.com/hashicorp/go-getter?tab=readme-ov-file#unarchiving
+		downloadURL = fmt.Sprintf("https://%s/releases/download/%s/%s",
+			e.Source, e.Version, enginePackageName(e))
+	}
+
+	// lock downloading process for only one instance
+	locks, err := downloadLocksFromContext(ctx)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	// locking by file where engine is downloaded
+	// however, it will not help in case of multiple parallel Terragrunt runs
+	locks.Lock(downloadFile)
```

Comment:
Shouldn't we check again to see if the file exists after the lock is acquired? The second routine that releases the lock should already have the engine downloaded, and shouldn't overwrite the original downloaded file.